### PR TITLE
luci-mod-status: add new channel analysis page

### PR DESCRIPTION
**Pull request by Ansuel**
Add channel analysis support. This can be very useful as a user can directly use the webui to check wifi channel utilization without using external tool. This use data already provided by iwinfo.

@jow- This is the """final""" implementation. If you are fine with this (and after some feedback with some testing), we still need to decide where to put this.  @hnyman advised to put this in a separate status menu. Currently this is added to the join wifi modal.

(some details about how the page works, a tab with all the radios is created. The pool function is unique and use the custom event 'tab-change' to detect the tab displayed and based on the tab name, updates the related graph and table. Multiple pool function were not acceptable since this would overload the wifi driver and also slowdown the wifi performance. Tab data is not cleared on tab switch and data is updated on the next pool update. A dummy wifi is added that rappresents the Local Interface.)

**Also this depends on the 4 patch I posted on the mailing list**
https://patchwork.ozlabs.org/project/openwrt/list/?series=218829 (v4)
They are rfc but they are already ready... just some concern in the comments.

Without the 4 patch, any wifi will be wrongly shown with 20mhz width.

Fixes: #4572
Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

### CODE REVIEW

**Comment by hnyman** [comment]
For some reason I can't get this to work in LuCI. The scan from LuCI hangs with spinning wheel on scan button, and Firefox shows a JS error:
image

But the channel widths are shown in the patches iwinfo itself.
 
```
root@router1:~# iwinfo wlan0 scan
Cell 01 - Address: 58:90:4xxxxx
          ESSID: "DNA-Wxxxx"
          Mode: Master  Channel: 36
          Signal: -90 dBm  Quality: 20/70
          Encryption: WPA2 PSK (CCMP)
          HT Operation:
                    Primary Channel: 36
                    Secondary Channel Offset: above
                    Channel Width: 20 MHz
          VHT Operation:
                    Channel Width: 80 MHz
                    Center Frequency 1: 42
                    Center Frequency 2: 0
```

(Not sure if I patched rpcd wrongly, or if that needs to be in the compiled image itself)

**Comment by Ansuel** [comment]
@hnyman yes the rpcd patch is needed to add the new data to ubus
the ui use this command to get the data
ubus call iwinfo scan '{"device":"wlan0"}'

Let me check the error with the function

**Comment by Ansuel** [comment]
Fixed the problem... (my fault working on old source....) pls give another test (or just change `callFrequencyList `to `this.callFrequencyList`)

**Comment by hnyman** [comment]
I recompiled everything and now it gives a slightly different error:
  "TypeError: this.callFrequencyList is not a function"

image

**Comment by hnyman** [comment]
But the scan itself seems to work:

```
root@router1:~# ubus call iwinfo scan '{"device":"wlan1"}'
{
        "results": [
                {
                        "ssid": "Koti_xxx",
                        "bssid": "FE:F5xxxx",
                        "mode": "Master",
                        "channel": 2,
                        "signal": -85,
                        "quality": 25,
                        "quality_max": 70,
                        "ht_operation": {
                                "primary_channel": 2,
                                "secondary_channel_offset": "above",
                                "channel_width": "any"
                        },
                        "encryption": {
                                "enabled": true,
                                "wpa": [
                                        1,
                                        2
                                ],
                                "authentication": [
                                        "psk"
                                ],
                                "ciphers": [
                                        "tkip",
                                        "ccmp"
                                ]
                        }
                },
                {
                        "ssid": "ASxxx",
                        "bssid": "30:85xxxx",
                        "mode": "Master",
                        "channel": 1,
                        "signal": -91,
                        "quality": 19,
                        "quality_max": 70,
                        "ht_operation": {
                                "primary_channel": 1,
                                "secondary_channel_offset": "no secondary",
                                "channel_width": "20 MHz"
 

```

**Comment by Ansuel** [comment]
Ok good @hnyman sorry again... Anyway to quickly test this... (if you don't want to recompile) Replace the js in the firmware with the one in this pr.

**Comment by hnyman** [comment]
Well, a different error...

image

Does the PR have the same code that actually works in your live router ????

**Comment by Ansuel** [comment]
Now yes... Ok... I placed the svg in the wrong path... Move channel_analysis.svg from` www/luci-static/resources/view` to `www/luci-static/resources`

**Comment by hnyman** [comment]
Well, now it actually worked ;-)

image

**Comment by Ansuel** [comment]
What about 5ghz?

**Comment by hnyman** [comment]
also 5 GHz works:

image

**Comment by Ansuel** [comment]
Thx a lot for the testing

File: modules/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js [hunk 1]
```diff
@@ -289,18 +290,157 @@ function add_dependency_permutations(o, deps) {
...
+				curr_offset += step;
+			}
+		}
+	}
+	createGraphHLine(G,curr_offset+step);
+}
+
+function add_wifi_to_graph(chan_analysis, res, scanCache) {
+	if (scanCache[res.bssid].color == null)
+		scanCache[res.bssid].color = '#'+('00000'+(Math.random()*(1<<24)|0).toString(16)).slice(-6);
```
Comment (jow-) [inline modules/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js:360]: I suggest to use the logic from [`firewall.getColorForName()`](https://github.com/openwrt/luci/blob/master/modules/luci-base/htdocs/luci-static/resources/firewall.js#L52) here, this ensures that the same network names always get the same pseudorandom colors.

Comment (Ansuel) [reply modules/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js]: To not have duplicated code, I generalized part of the `firewall.getColorForName()` and added to prng.

**Comment by Ansuel** [comment]
Added some small modification. 
- line and text are now in a group 
- added support for stale wifi (opacity for wifi detected as stale)
- added a point before the wifi to quickly identify the wifi by the color (useful with wifi with the same name)

Immagine 2020-11-18 024410

**Comment by pkgadd** [comment]
I've successfully tested the current version (b3505b7) on ipq8065/ qca9984 (nbg6817) and lantiq/ AR9227+QCA9880 (bthub5a), it's a nice improvement. 

For the graphical representation it might be even better to inject the local AP interfaces (of the device running the test) into the channel utilization as well.

**Comment by hnyman** [comment]
Interesting, the modified color randomiser now causes an error for me:

image

Apparently it worked for @pkgadd so I wonder what happens here.

**Comment by Ansuel** [comment]
@hnyman think you missed the other commit

**Comment by hnyman** [comment]
I didn't.
It was a browser cache problem.

Apparently the .js file did not get marked as new, as I did not commit the changes, so the version string remained static "prng.js?v=git-20.311.85590-5c5b134". 

Clearing browser cache and restarting the browser helped ;-)

**Comment by Ansuel** [comment]
(shift+refresh = load the page + purge cache for that site)

**Comment by hnyman** [comment]
That didn't work.  Tried it first...

But in any case, the enw code seems to work.

**Comment by Ansuel** [comment]
Do you have any way to test if 5 ghz 160 mhz or 80+80 band works?

**Comment by hnyman** [comment]
Not really.

**Comment by pkgadd** [comment]
Sorry, I don't have any 160 MHz or 80+80 MHz channels in my environment either.

**Comment by Ansuel** [comment]
I tested 80+80 and 160 by manually setting the data. Looks to work so i think this is ready... We still need to decide the place of this. @jow- any hint on that? Is it good? (graph in the wifi scan button?)

**Comment by hnyman** [comment]
That works ok, and it could be also there in any case.
But people rarely need to do "scan for joining", so the nice feature may remain hidden there.

I would also add it to the status menu

**Comment by Ansuel** [comment]
Ok so i will change this to
1. move the graph in a separate page
2. parse the wifi data differently and show more data related to channel width
3. add a button to the wifi settings that redirect to the channel analysis page

**Comment by Ansuel** [comment]
@hnyman I moved the feature to a new page. What do you think?

File: modules/luci-base/htdocs/luci-static/resources/ui.js [hunk 2]
```diff
@@ -3597,9 +3597,11 @@ var UI = baseclass.extend(/** @lends LuCI.ui.prototype */ {
 				this.setActiveTabId(panes[selected], selected);
 			}
 
-			panes[selected].dispatchEvent(new CustomEvent('cbi-tab-active', {
-				detail: { tab: panes[selected].getAttribute('data-tab') }
-			}));
+			requestAnimationFrame(L.bind(function(pane) {
```
Comment (hnyman) [inline modules/luci-base/htdocs/luci-static/resources/ui.js:3600]: nitpicking, but there is a space at the end of the line.
"whitespace error"

**Comment by hnyman** [comment]
Thanks. I like the new "Channel Analysis" page in the status menu.

Merging this requires that @jow- first commits the needed patches for iwinfo and rpcd (and the needed version bumps in OpenWrt repo):
https://patchwork.ozlabs.org/project/openwrt/list/?series=214580

Ps. note that your second patch there patches "rpcd" while the commit title says "iwinfo"...

**Comment by Ansuel** [comment]
Yes I'm following the naming scheme. If you notice in the rpcd repo,
changes related to iwinfo rpcd module have iwinfo as title.

Il Sab 21 Nov 2020, 09:12 Hannu Nyman <notifications@github.com> ha scritto:

**Comment by Ansuel** [comment]
@hnyman i'm thinking of adding to the graph but i need to confirm something...
Do you have a wifi that has set 40mhz width but is actually working to 20mhz? 

I need to check if ht_mode in `ubus call iwinfo info '{"device":"radio1"}'` is what the wifi is configured or what the wifi is actually using. (40 mhz set but due to reg and other wifi is using 20 mhz)

**Comment by hnyman** [comment]
> check if ht_mode in ubus call iwinfo info '{"device":"radio1"}' is what the wifi is configured or what the wifi is actually using. (40 mhz set but due to reg and other wifi is using 20 mhz)

You mean, the width for the radio in the device itself ???

I think that is the configured one, not the really running channel width. Sadly.
Not quite sure, but might be so.

I tried to test that with VHT160 by setting it to 160 and then doing a scan from my other router:
* device itself says: htmode 160
* scan from other device says:  80

Device itself:
```
root@router1:/etc/config# ubus call iwinfo info '{"device":"radio0"}'
{
        "phy": "phy0",
        "bssid": "DC:....",
        "country": "FI",
        "mode": "Master",
        "channel": 36,
        "frequency": 5180,
        "frequency_offset": 0,
        "txpower": 23,
        "txpower_offset": 0,
        "quality_max": 70,
        "noise": 0,
        "htmodes": [
                "HT20",
                "HT40",
                "VHT20",
                "VHT40",
                "VHT80",
                "VHT80+80",
                "VHT160"
        ],
        "hwmodes": [
                "ac",
                "n"
        ],
        "hwmode": "ac",
        "htmode": "VHT160",
        "hardware": {
                "id": [
                        5772,
                        70,
                        5772,
                        51966
                ],
                "name": "Qualcomm Atheros QCA9984"
        }
}
```

Seen from other router:
image

Testing with VHT160 is maybe not quite reliable, but still hints that it may be misleading.

**Comment by Ansuel** [comment]
Will check if I can extract the info from iwinfo... (the center_chan entry should be what we actually use)

**Comment by Ansuel** [comment]
Any I'm stupid and I never check my spelling... My fault I do this kind of thing in the night. Anyway
The new patch
https://patchwork.ozlabs.org/project/openwrt/list/?series=216286

@hnyman would really appreciate some comments about the 'Local Interface' wifi...
- I set it to a dummy -10 signal (doesn't make sense to have a signal of a station)
- The new patchset is needed to know the right channel to parse

**Comment by pkgadd** [comment]
I really love those changes, works nicely on my nbg6817 and bthub5:

2.4 GHz
2g

5 GHz
5g

Tested c89eed8d1e5ba2dc9369caf86d6332e474a64712 and v2 of your rpcd/ iwinfo patches on top of r15021-b0ecae504b and kernel v5.4.79.

**Comment by pkgadd** [comment]
It works also nicely on a tri-radio device (ipq4019/ map-ac2200) with segmented channel coverage (radio0 ch36-ch64, radio2 ch100-ch165).

**Comment by openwrt-diy** [comment]
This feature is great, can it be applied to OpenWrt-19.07?

**Comment by hnyman** [comment]
@jow- 
I think that this works pretty nicely.

Can you look at evaluating/committing the necessary patches to iwinfo and rpcd, which ansuel submitted to mailing list adn patchwork.

image

@Ansuel 
This is past WIP quality, I think.  ;-)

**Comment by Ansuel** [comment]
Still dubious about the Local Interface data and naming!
Also thx to @pkgadd i need to fix the `20 or 40mhz` bug

**Comment by Ansuel** [comment]
Added a way to check when a wifi has the Intolerant Bit set (for 40MHz channel) and fixed the problem :D

**Comment by pkgadd** [comment]
Just to confirm, I've successfully tested this version as well (ipq8065, ipq4019, lantiq; all ath9k/ ath10k wireless).
`Tested-by: Stefan Lippers-Hollmann <s.l-h@gmx.de>`

**Comment by hnyman** [comment]
Works for me, too.

This LuCI PR could be merged once @jow- (or somebody else with write access to OpenWrt main repo) merges the needed changes to rpcd and iwinfo.

**Comment by jow-** [comment]
Unfortunately I can't merge the rpcd and iwinfo patches as is. The iwinfo changes in particular will break the existing ABI. I also do not the like the way the JSON information is structured (e.g. using "40 Mhz" as field value instead of something like `{ "mhz": 40 }`.

I agree that this feature is nice, but the iwinfo side needs more polishing first.

**Comment by Ansuel** [comment]
Sure I can make all the needed change. About the ABI there is no way to resolve the problem?

**Comment by Ansuel** [comment]
@jow- sorry for the tag. Is there a way to add the iwinfo changes with the ABI problem? I could be very wrong but the patches add data and add entries at the end of the struct so in theory bin compiled on the old library should still work.

**Comment by hnyman** [comment]
I think that you need to properly bump the ABI version in Makefile.

https://github.com/openwrt/openwrt/blob/4dc5b647602dc33be9bc940e91e5b849fdc98015/package/network/utils/iwinfo/Makefile#L36

**Comment by Ansuel** [comment]
Think that bump should be done by jow on iwinfo update on the openwrt repo. Think that jow was referring to the fact that he wanted to avoid in any way possibile an ABI change.

**Comment by jow-** [comment]
There's multiple things that I'd like to see done on the iwinfo side:
 - change the libiwinfo packaging to ship a versioned shared library (libiwinfo.so.$ABI_VERSION)
 - bump the ABI version
 - change the JSON output format, to produce something like `"channel_width": 40` or `"channel_width": 8080` instead of `"40 Mhz" or "80 + 80 MHz" respectively
 - add appropriate stubs for `struct iwinfo_ops.center_chan1` and `struct iwinfo_ops.center_chan2` to the other backends *or* check whether `ops->center_chan1` is != NULL before calling it. Right now, the code will segfault with non-nl80211 backends.

**Comment by Ansuel** [comment]
> There's multiple things that I'd like to see done on the iwinfo side:
> 
> * change the libiwinfo packaging to ship a versioned shared library (libiwinfo.so.$ABI_VERSION)

My statement could look a bit rude but I think it would be better if you or someone more skilled than me do this change. I don't want to create some disaster.

> * bump the ABI version

Same thing. (since the iwinfo makefile in the openwrt repo has a specific commit, we can work on fixing the proposed patch and then work on the changes needed to the makefile.

> * change the JSON output format, to produce something like `"channel_width": 40` or `"channel_width": 8080` instead of `"40 Mhz"` or `"80 + 80 MHz"` respectively - didn't look in detail but if multiple things are possible then an array would work, e.g. `"channel_width": [ 20, 40, ... ]`

The width is unique. So an enum that store these value should be sufficent

> * add appropriate stubs for `struct iwinfo_ops.center_chan1` and `struct iwinfo_ops.center_chan2` to the other backends _or_ check whether `ops->center_chan1` is != NULL before calling it. Right now, the code will segfault with non-nl80211 backends.

Think the check is better than a stub that would return 0 or -1. What do you think?

**Comment by jow-** [comment]
> Think the check is better than a stub that would return 0 or -1. What do you think?

Yes, a check is fine with me.

**Comment by Ansuel** [comment]
I pushed v4 https://patchwork.ozlabs.org/project/openwrt/list/?series=218829 and updated the pr to reflect the changes.
Can someone test?

**Comment by hnyman** [comment]
> Can someone test?

v4 patch series may have a whitespace error.  Spaces instead of tab on one original line.

image

(similar spaces are apparently also on the lines below, but those spaces on added lines do not break patching.)


Fixing that tab in the local patch helped me:
```
--- a/package/network/utils/iwinfo/patches/patch1.patch
+++ b/package/network/utils/iwinfo/patches/patch1.patch
@@ -81,7 +81,7 @@ index 0332bc2..6bfe0ab 100644
                        format_quality_max(e->quality_max));
 -              printf("          Encryption: %s\n\n",
 +              printf("          Encryption: %s\n",
-                      format_encryption(&e->crypto));
+                       format_encryption(&e->crypto));
 +              printf("          HT Operation:\n");
 +              printf("                    Primary Channel: %d\n",
 +                     e->ht_chan_info.primary_chan);

```

I will come back to feedback on the functionality itself, once I get it compiled ;-)

**Comment by hnyman** [comment]
After fixing that whitespace error in the v4 iwinfo patch, the channel analysis functionality itself works ok ;-)

image

**Comment by hnyman** [comment]
Hmm, I noticed one minor issue.
I am not quite sure of the exact reason but this is my speculative guess:

If there is a weak station that disappears during a periodic refresh scan, it is taken away from the list, but it may stay on the graph.

See the "kotiboksi..." here. It was shown both on the graph and the list on the first scan, but disappeared from the list in an automatic refresh, maybe 5 seconds later (or is it 10 seconds, I haven't looked). But it stayed visible in the graph.

image

Not a major issue,  but might be something to be fixed later.

**Comment by Ansuel** [comment]
@hnyman 'It's not a bug, it's a feature' cit.
In theory the page should show the channel utilization so any station that comes up and disappear after a while should be still displayed as it actually exists and cause some interference. Do you think this is wrong?
(about the whitespace errors, in theory they should be there to correctly align the args of the function at the next line. Does the patch complains about them? Could be that I messed up that 2 line)

**Comment by hnyman** [comment]
> Does the patch complains about them? Could be that I messed up that 2 line)

Yes it does. The build breaks.
Thatswhy I flagged the issue.

**Comment by hnyman** [comment]
The iwinfo & rpcd part were committed yesterday, so I think that this LuCI part could also be merged.

Ps.
I think that @jow- 's comments yesterday on mailing list regarding iwinfo and rpcd changes were not yet noted when @dangowrt committed your earlier changes, so I suspect that a new modification round for rpcd and iwinfo might be needed. The old version is now live in OpenWrt master.

**Comment by Ansuel** [comment]
Mhhh yep but I think I have to push a new patch to address jow changes.

Il Mer 6 Gen 2021, 08:58 Hannu Nyman <notifications@github.com> ha scritto:

**Comment by Ansuel** [comment]
Anyway I'm really happy that this feature finally will be added. Honestly it's a very useful feature as the very base of the OpenWRT project is Wireless Freedom.

**Comment by Ansuel** [comment]
@hnyman the required commits are now upstream. We just need to wait a review from jow and this can be merged.

**Comment by hnyman** [comment]
>  the required commits are now upstream. We just need to wait a review from jow

Jow made additions/changes to the iwinfo Makefile and merged the bump there, so I think that this LuCI part can now be also merged.

The current LuCI commit version has been in my own builds for the last month and has worked ok in several routers. So lets just merge this and wait for possible feedback (and/or change tweaks).

**Comment by dangowrt** [comment]
@hnyman @Ansuel have you tried LuCI after this got merged? I'm kinda stuck with non-working iwinfo, symptoms feel like libiwinfo-lua is missing (it's not, ldd shows all ok) but since everything switched to JavaScript I have absolutely no clue at all what's going on and how it is supposed to work.

**Comment by Ansuel** [comment]
Honestly in past test... I encounter some problem with these patch if the center chan were at the center of the struct instead of putting them at the bottom. But it should be caused by versioning and not using ABI. 
anyway this use js and ubus so lua is completely skipped.

**Comment by hnyman** [comment]
I have used the channel analysis page since November. The last LuCI patch in this PR stayed the same since 6 Dec and has been included in all my builds.  It has worked ok, and works even currently.

I have currently it in R7800 with  iwinfo/libiwinfo/-lua 2021-01-06-23d2722b-1 with LuCI luci-mod-status git-20.340.80217-277ffef  manually patched with an identical patch as this merged LuCI PR commit.
* The channel analysis works.
* But the other iwinfo related stuff seems semi-dead. (wireless status info in overview shows up almost empty, and in network/wireless is shows as as not associated. But associated devices are still shown with sensible info...)

As this LuCI channel analysis patch has not changed and has no impact on those other LuCI pages, it is likely related to the last minute non-LuCI changes to the iwinfo & rpcd stuff. 

The other wireless packages have worked ok until 5 January (with the patches then available in the mailing list).

My first guess is that the last minute change on January 6th have somehow changed the iwinfo output, so that the other LuCI components need now a refresh.

I have actually in my R7800 archive several firmware versions from 6th and 7th, e.g. one with r15437-b4c0d377f6-20210106 meaning right after the first version of iwinfo & rpcd were merged. I will re-flash those and investigate a bit, when the breakage happens.

**Comment by jow-** [comment]
Fixed by bc121a5a456b564618423b67a6aef859c0ba6a87

**Comment by BashfulBladder** [comment]
@hnyman can you elaborate on this:

> But the other iwinfo related stuff seems semi-dead. (wireless status info in overview shows up almost empty, and in network/wireless is shows as as not associated. But associated devices are still shown with sensible info...)

I have 2 R700 (RC3 & another some NSS build) that exhibit:

- days of only localhost in 5GHz, 2.4GHz had a 100 cells (both in iw & iwinfo)
- noise absent on 2.4GHz (sort of expected as it doesn't have a network configured)
- noise present only in 2 channels in 5GHz (both in iw & iwinfo using this patch below)

```
--- a/iwinfo_cli.c	2021-01-31 09:05:58.000000000 -0500
+++ b/iwinfo_cli.c	2021-07-31 11:45:25.000000000 -0400
@@ -677,6 +677,32 @@
 }
 
 
+static void print_channel_survey(const struct iwinfo_ops *iw, const char *ifname)
+{
+	int i, len;
+	char buf[IWINFO_BUFSIZE];
+	struct iwinfo_survey_entry *e;
+
+	if (iw->survey(ifname, buf, &len) || len <= 0)
+	{
+		printf("No channel survey information available\n");
+		return;
+	}
+
+	for (i = 0; i < len; i += sizeof(struct iwinfo_survey_entry))
+	{
+		e = (struct iwinfo_survey_entry *) &buf[i];
+		
+		printf("Survey data from %s frequency %d MHz: %s noise (active time: %llu ms, busy time: %llu ms)\n",
+			ifname,
+			e->mhz,
+			format_noise((int8_t)e->noise),
+			e->active_time,
+			e->busy_time);			
+	}
+}
+
+
 static void print_txpwrlist(const struct iwinfo_ops *iw, const char *ifname)
 {
 	int len, pwr, off, i;
@@ -878,6 +904,7 @@
 			"	iwinfo <device> assoclist\n"
 			"	iwinfo <device> countrylist\n"
 			"	iwinfo <device> htmodelist\n"
+			"	iwinfo <network> noiselist\n"
 			"	iwinfo <backend> phyname <section>\n"
 		);
 
@@ -973,6 +1000,10 @@
 				case 'h':
 					print_htmodelist(iw, argv[1]);
 					break;
+				
+				case 'n':
+					print_channel_survey(iw, argv[1]);
+					break;
 
 				default:
 					fprintf(stderr, "Unknown command: %s\n", argv[i]);
```
I can get a richer result if I employ:

```
ifconfig wlan0 down
iw phy phy0 interface add scan0 type station
ifconfig scan0 up
iwinfo scan0 noiselist		(< nope, not yet)
iwinfo scan0 scan		(<- this yields 30 cells AND noise across all channels)
iwinfo scan0 noiselist
iw dev scan0 del
ifconfig wlan0 up
killall -HUP hostapd
```
And sure enough it remains that way until the next reboot. Still, that doesn't explain why it it needs to happen at all after a reboot.

**Comment by Neustradamus** [comment]
@hnyman, @abajk: Have you seen the @BashfulBladder comment?

**Pull request merged**